### ISSUE TITLE
SignalR doesn't always reliable detect TCP/IP issues

### ISSUE THREAD

**Issue by dannyhaak** [dotnet/aspnetcore#43392]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

Use SignalR between Linux embedded devices in the field and our server application. In general it is very reliable, except when the connection is really bad. And this can mean different things: switching between uploads, saturated uploads, etc. 

I am now able to reproduce this on my desk. The embedded linux device is in a switch, and I just unplug the switch when the connection is active. Due to TCP/IP time-outs, it takes a while before SignalR detects this. The verbose logging I get is something like:

```
[15:48:21.717 INF WebSocketsTransport] Transport is stopping.
<nothing, waited for 10 minutes>
```

And that means that the `connection.Closed` handler is simply never called, making it impossible for us to properly reconnect.

Connection set-up:
```
connection = new HubConnectionBuilder()
                .WithUrl(uriWithQueryString)
                .AddMessagePackProtocol(options =>
                {
                    options.SerializerOptions = MessagePackSerializerOptions.Standard
                        .WithResolver(ContractlessStandardResolver.Instance);
                })
                .ConfigureLogging(logging =>
                {
                    logging.SetMinimumLevel(LogLevel.Debug);
                    logging.AddSerilog();
                })
                .Build();
            
            connection.Closed += ConnectionClosed;

            connection.On<SettingsRoot>("UpdateSettings", ReceivedNewSettings);
            connection.On<ReadPointTrigger>("Trigger", ReceivedTrigger);
```

### Expected Behavior

This is what I expect in the logging, and that happens sometimes:

```
[15:47:20.449 INF WebSocketsTransport] Transport is stopping.
[15:47:21.049 INF HttpConnection] HttpConnection Disposed.
[15:47:21.470 ERR HubConnection] Connection is shutting down due to an error.
System.TimeoutException: Server timeout (30000.00ms) elapsed without receiving a message from the server.
[15:47:27.705 INF BackendService] Connection closed: Server timeout (30000.00ms) elapsed without receiving a message from the server.
```

(The last log line is indicating that my connect.Closed handler has been called).

### Steps To Reproduce

I can reliable reproduce this on my local network. It doesn't happen always, but quite some times.

### .NET Version

6.0.400

### Anything else?

I tested with the SignalR.Client 6 series and 7.0.0-preview-7, and both show the same behaviour.

Linux version of our app is build using:
`dotnet publish -r linux-arm x.csproj -c Release --self-contained`

```
rfidadm@FX75007FF746:~$ uname -r 
4.9.182-rt131-yocto-standard
```

**Comment by BrennanConroy** [comment]
Turn on Debug or Trace logs please. When the WebSocket transport is closing it waits to clean up the send and receive loops and those should have some logs.

**Comment by davidfowl** [comment]
To set expectations, this area will never be as reliable as developers expect, it's hard to reproduce and hard to get right. Networking is complex and inconsistent. That said, we're constantly writing to the transport and looking for pings to detect bad connections so I'm surprised to takes 10 minutes to detect a bad connection. Since you can reproduce reliably, then you can help us debug (welcome to the team 😄).

**Comment by dannyhaak** [comment]
Thanks for your quick reply. Let me try to help with some more logging:

Situation after I unplugged the Ethernet cable:
```
[10:28:01.827 VRB HubConnection] Sending item for stream '4'.
[10:28:02.200 VRB HubConnection] Waiting on Connection Lock in SendStreamItems (/_/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:2077).
[10:28:02.206 DBG HubConnection] Sending StreamItemMessage message '3'.
[10:28:02.210 DBG HubConnection] Sending StreamItemMessage message '3' completed.
[10:28:02.214 VRB HubConnection] Releasing Connection Lock in SendWithLock (/_/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:957).
[10:28:02.223 VRB HubConnection] Sending item for stream '3'.
[10:28:02.226 VRB HubConnection] Waiting on Connection Lock in SendStreamItems (/_/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:2077).
[10:28:02.231 DBG HubConnection] Sending StreamItemMessage message '3'.
[10:28:02.235 DBG HubConnection] Sending StreamItemMessage message '3' completed.
[10:28:02.239 VRB HubConnection] Releasing Connection Lock in SendWithLock (/_/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:957).
[10:28:02.245 VRB HubConnection] Sending item for stream '3'.
[10:28:02.798 VRB HubConnection] Waiting on Connection Lock in SendStreamItems (/_/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:2077).
[10:28:02.806 DBG HubConnection] Sending StreamItemMessage message '4'.
[10:28:02.810 DBG HubConnection] Sending StreamItemMessage message '4' completed.
[10:28:02.818 VRB HubConnection] Releasing Connection Lock in SendWithLock (/_/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:957).
[10:28:02.824 VRB HubConnection] Sending item for stream '4'.
[10:28:03.798 VRB HubConnection] Waiting on Connection Lock in SendStreamItems (/_/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:2077).
[10:28:03.811 DBG HubConnection] Sending StreamItemMessage message '4'.
[10:28:03.825 DBG HubConnection] Sending StreamItemMessage message '4' completed.
[10:28:03.829 VRB HubConnection] Releasing Connection Lock in SendWithLock (/_/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:957).
[10:28:03.837 VRB HubConnection] Sending item for stream '4'.
[10:28:04.146 VRB HubConnection] Waiting on Connection Lock in HandleConnectionClose (/_/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:1362).
[10:28:04.192 DBG HttpConnection] Disposing HttpConnection.
[10:28:04.220 INF WebSocketsTransport] Transport is stopping.
[10:28:04.354 VRB HubConnection] Stream '4' has been canceled by client.
[10:28:04.360 VRB HubConnection] Sending completion message for stream '4'.
[10:28:04.370 VRB HubConnection] Waiting on Connection Lock in CommonStreaming (/_/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:2077).
[10:28:04.382 VRB HubConnection] Stream '3' has been canceled by client.
[10:28:04.386 VRB HubConnection] Sending completion message for stream '3'.
[10:28:04.389 VRB HubConnection] Waiting on Connection Lock in CommonStreaming (/_/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:2077).
[10:28:04.398 VRB HubConnection] Stream '2' has been canceled by client.
[10:28:04.401 VRB HubConnection] Sending completion message for stream '2'.
[10:28:04.403 VRB HubConnection] Waiting on Connection Lock in CommonStreaming (/_/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:2077).
```

And then no further logging here, for like forever.

Sometimes, when you put back in the Ethernet cable relatively quickly, it will restore automatically. Below logging only appears when you put the cable back in.

```
[10:29:33.464 DBG WebSocketsTransport] Receive loop stopped.
[10:29:33.502 ERR WebSocketsTransport] Error while sending a message.
System.OperationCanceledException: The operation was canceled.
 ---> System.IO.IOException: Unable to write data to the transport connection: Broken pipe.
 ---> System.Net.Sockets.SocketException (32): Broken pipe
   --- End of inner exception stack trace ---
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.ThrowException(SocketError error, CancellationToken cancellationToken)
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.System.Threading.Tasks.Sources.IValueTaskSource.GetResult(Int16 token)
   at System.Net.Security.SslStream.<WriteSingleChunk>g__CompleteWriteAsync|182_1[TIOAdapter](ValueTask writeTask, Byte[] bufferToReturn)
   at System.Net.Security.SslStream.WriteAsyncInternal[TIOAdapter](TIOAdapter writeAdapter, ReadOnlyMemory`1 buffer)
   at System.Net.Http.HttpConnection.RawConnectionStream.WaitWithConnectionCancellationAsync(ValueTask task, HttpConnection connection, CancellationToken cancellationToken)
   at System.Net.WebSockets.ManagedWebSocket.WaitForWriteTaskAsync(ValueTask writeTask)
   --- End of inner exception stack trace ---
   at System.Net.WebSockets.ManagedWebSocket.WaitForWriteTaskAsync(ValueTask writeTask)
   at Microsoft.AspNetCore.Http.Connections.Client.Internal.WebSocketsTransport.StartSending(WebSocket socket)
[10:29:34.067 DBG WebSocketsTransport] Send loop stopped.
[10:29:34.130 DBG WebSocketsTransport] Transport stopped.
[10:29:34.189 INF HttpConnection] HttpConnection Disposed.
[10:29:34.383 DBG HubConnection] Canceling all outstanding invocations.
[10:29:34.391 ERR HubConnection] Connection is shutting down due to an error.
System.TimeoutException: Server timeout (30000.00ms) elapsed without receiving a message from the server.
[10:29:34.726 VRB HubConnection] The HubConnection is attempting to transition from the Connected state to the Disconnected state.
[10:29:34.895 VRB HubConnection] Releasing Connection Lock in HandleConnectionClose (/_/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:1396).
[10:29:34.939 DBG HubConnection] Invoking the Closed event handler.
[10:29:35.004 INF BackendService] Connection closed: Server timeout (30000.00ms) elapsed without receiving a message from the server.
[10:29:35.012 VRB HubConnection] Releasing Connection Lock in CommonStreaming (/_/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:2081).
[10:29:35.054 VRB HubConnection] Releasing Connection Lock in CommonStreaming (/_/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:2081).
[10:29:35.085 VRB HubConnection] Releasing Connection Lock in CommonStreaming (/_/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:2081).
```

**Comment by dannyhaak** [comment]
Another situation:

```
[10:33:05.148 VRB HubConnection] Releasing Connection Lock in SendWithLock (/_/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:957).
[10:33:05.152 VRB HubConnection] Sending item for stream '3'.
[10:33:06.092 VRB HubConnection] Waiting on Connection Lock in SendStreamItems (/_/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:2077).
[10:33:06.103 DBG HubConnection] Sending StreamItemMessage message '4'.
[10:33:06.112 DBG HubConnection] Sending StreamItemMessage message '4' completed.
[10:33:06.115 VRB HubConnection] Releasing Connection Lock in SendWithLock (/_/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:957).
[10:33:06.118 VRB HubConnection] Sending item for stream '4'.
[10:33:07.092 VRB HubConnection] Waiting on Connection Lock in SendStreamItems (/_/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:2077).
[10:33:07.097 DBG HubConnection] Sending StreamItemMessage message '4'.
[10:33:07.101 DBG HubConnection] Sending StreamItemMessage message '4' completed.
[10:33:07.104 VRB HubConnection] Releasing Connection Lock in SendWithLock (/_/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:957).
[10:33:07.107 VRB HubConnection] Sending item for stream '4'.
[10:33:08.096 VRB HubConnection] Waiting on Connection Lock in SendStreamItems (/_/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:2077).
[10:33:08.101 DBG HubConnection] Sending StreamItemMessage message '4'.
[10:33:08.114 DBG HubConnection] Sending StreamItemMessage message '4' completed.
[10:33:08.119 VRB HubConnection] Releasing Connection Lock in SendWithLock (/_/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:957).
[10:33:08.125 VRB HubConnection] Sending item for stream '4'.
[10:33:08.629 VRB HubConnection] Waiting on Connection Lock in SendStreamItems (/_/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:2077).
[10:33:08.633 DBG HubConnection] Sending StreamItemMessage message '3'.
[10:33:08.638 DBG HubConnection] Sending StreamItemMessage message '3' completed.
[10:33:08.643 VRB HubConnection] Releasing Connection Lock in SendWithLock (/_/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:957).
[10:33:08.647 VRB HubConnection] Sending item for stream '3'.
[10:33:09.099 VRB HubConnection] Waiting on Connection Lock in SendStreamItems (/_/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:2077).
[10:33:09.104 DBG HubConnection] Sending StreamItemMessage message '4'.
[10:33:09.113 DBG HubConnection] Sending StreamItemMessage message '4' completed.
[10:33:09.119 VRB HubConnection] Releasing Connection Lock in SendWithLock (/_/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:957).
[10:33:09.122 VRB HubConnection] Sending item for stream '4'.
[10:33:10.100 VRB HubConnection] Waiting on Connection Lock in SendStreamItems (/_/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:2077).
[10:33:10.105 DBG HubConnection] Sending StreamItemMessage message '4'.
[10:33:10.109 DBG HubConnection] Sending StreamItemMessage message '4' completed.
[10:33:10.118 VRB HubConnection] Releasing Connection Lock in SendWithLock (/_/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:957).
[10:33:10.121 VRB HubConnection] Sending item for stream '4'.
[10:33:10.725 VRB HubConnection] Waiting on Connection Lock in HandleConnectionClose (/_/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:1362).
[10:33:10.731 VRB HubConnection] Stream '2' has been canceled by client.
[10:33:10.743 VRB HubConnection] Sending completion message for stream '2'.
[10:33:10.740 VRB HubConnection] Stream '4' has been canceled by client.
[10:33:10.750 VRB HubConnection] Sending completion message for stream '4'.
[10:33:10.735 VRB HubConnection] Stream '3' has been canceled by client.
[10:33:10.746 DBG HttpConnection] Disposing HttpConnection.
[10:33:10.755 INF WebSocketsTransport] Transport is stopping.
[10:33:10.756 VRB HubConnection] Waiting on Connection Lock in CommonStreaming (/_/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:2077).
[10:33:10.754 VRB HubConnection] Sending completion message for stream '3'.
[10:33:10.749 VRB HubConnection] Waiting on Connection Lock in CommonStreaming (/_/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:2077).
[10:33:10.763 VRB HubConnection] Waiting on Connection Lock in CommonStreaming (/_/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:2077).
[10:33:12.487 DBG BackendService] Current cpuUsage 42.2 % and memoryUsage 79757312
[10:33:42.496 DBG BackendService] Current cpuUsage 17.7 % and memoryUsage 79712256
[10:34:12.510 DBG BackendService] Current cpuUsage 16.0 % and memoryUsage 79671296
[10:34:42.503 DBG BackendService] Current cpuUsage 17.3 % and memoryUsage 78884864
[10:35:12.489 DBG BackendService] Current cpuUsage 18.3 % and memoryUsage 77381632
[10:35:42.499 DBG BackendService] Current cpuUsage 14.1 % and memoryUsage 77361152
[10:36:12.493 DBG BackendService] Current cpuUsage 15.4 % and memoryUsage 77373440
[10:36:42.497 DBG BackendService] Current cpuUsage 14.3 % and memoryUsage 77352960
[10:37:12.507 DBG BackendService] Current cpuUsage 15.8 % and memoryUsage 77357056
```

Here I waited more than two minutes (longer than the HttpClient timeout?) before putting the cable back in, and then the connection doesn't recover within the next 10 minutes, nor do we get a disconnect notification. This is the most problematic situation for us.

**Comment by dannyhaak** [comment]
I also tried to get it to reproduce on my desktop Mac M1, but to no avail. It's good to know that the Linux machine this code is running on is slow (dual core 600 MHz), and under high CPU load (~50%).

Log from proper working set-up on my Mac M1; it seems that the Connection Lock is the culprit in the above examples?

```
[10:50:33.965 DBG WebSocketsTransport] Received message from application. Payload size: 497.
[10:50:34.618 VRB HubConnection] Waiting on Connection Lock in SendStreamItems (/_/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:2077).
[10:50:34.619 DBG HubConnection] Sending StreamItemMessage message '3'.
[10:50:34.619 DBG HubConnection] Sending StreamItemMessage message '3' completed.
[10:50:34.619 VRB HubConnection] Releasing Connection Lock in SendWithLock (/_/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:957).
[10:50:34.619 VRB HubConnection] Sending item for stream '3'.
[10:50:34.621 DBG WebSocketsTransport] Received message from application. Payload size: 389.
[10:50:34.966 VRB HubConnection] Waiting on Connection Lock in SendStreamItems (/_/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:2077).
[10:50:34.966 DBG HubConnection] Sending StreamItemMessage message '4'.
[10:50:34.966 DBG HubConnection] Sending StreamItemMessage message '4' completed.
[10:50:34.966 VRB HubConnection] Releasing Connection Lock in SendWithLock (/_/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:957).
[10:50:34.966 VRB HubConnection] Sending item for stream '4'.
[10:50:34.966 DBG WebSocketsTransport] Received message from application. Payload size: 489.
[10:50:35.117 VRB HubConnection] Waiting on Connection Lock in HandleConnectionClose (/_/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:1362).
[10:50:35.117 DBG HttpConnection] Disposing HttpConnection.
[10:50:35.117 INF WebSocketsTransport] Transport is stopping.
[10:50:35.117 VRB HubConnection] Stream '4' has been canceled by client.
[10:50:35.121 VRB HubConnection] Sending completion message for stream '4'.
[10:50:35.121 VRB HubConnection] Waiting on Connection Lock in CommonStreaming (/_/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:2077).
[10:50:35.118 VRB HubConnection] Stream '3' has been canceled by client.
[10:50:35.123 VRB HubConnection] Sending completion message for stream '3'.
[10:50:35.124 VRB HubConnection] Waiting on Connection Lock in CommonStreaming (/_/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:2077).
[10:50:35.118 VRB HubConnection] Stream '2' has been canceled by client.
[10:50:35.124 VRB HubConnection] Sending completion message for stream '2'.
[10:50:35.124 VRB HubConnection] Waiting on Connection Lock in CommonStreaming (/_/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:2077).
[10:50:35.121 DBG WebSocketsTransport] Send loop stopped.
[10:50:35.126 DBG WebSocketsTransport] Transport stopped.
[10:50:35.127 INF HttpConnection] HttpConnection Disposed.
[10:50:35.131 DBG HubConnection] Canceling all outstanding invocations.
[10:50:35.131 ERR HubConnection] Connection is shutting down due to an error.
System.TimeoutException: Server timeout (30000,00ms) elapsed without receiving a message from the server.
[10:50:35.133 DBG WebSocketsTransport] Receive loop canceled.
[10:50:35.135 DBG WebSocketsTransport] Receive loop stopped.
[10:50:35.136 VRB HubConnection] The HubConnection is attempting to transition from the Connected state to the Disconnected state.
[10:50:35.136 VRB HubConnection] Releasing Connection Lock in HandleConnectionClose (/_/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:1396).
[10:50:35.137 DBG HubConnection] Invoking the Closed event handler.
[10:50:35.137 VRB HubConnection] Releasing Connection Lock in CommonStreaming (/_/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:2081).
[10:50:35.137 INF BackendService] Connection closed: Server timeout (30000,00ms) elapsed without receiving a message from the server.
```

**Comment by BrennanConroy** [comment]
I think I can see an issue with the WebSocketsTransport, would it be possible for you to try an experiment where you replace the WebSocket implementation with a custom one below and when you trigger the issue (and wait more than 30 seconds) you cancel the token (also shown below)?

```csharp
var cts = new CancellationTokenSource();

new HubConnectionBuilder()
    .WithUrl(uriWithQueryString, options =>
    {
        options.WebSocketFactory = async (context, _) =>
        {
            var websocket = new WebSocketWrapper(new ClientWebSocket(), cts.Token);
            await websocket.ConnectAsync(context.Uri, CancellationToken.None);
            return websocket;
        };
    })
    ...

// after unplugging ethernet and waiting at least 30 seconds
cts.Cancel();

internal class WebSocketWrapper : WebSocket
{
    private readonly ClientWebSocket _inner;
    private readonly CancellationToken _cancellationToken;

    public WebSocketWrapper(ClientWebSocket inner, CancellationToken cancellationToken)
    {
        _inner = inner;
        _cancellationToken = cancellationToken;
    }

    public Task ConnectAsync(Uri uri, CancellationToken cancellationToken) => _inner.ConnectAsync(uri, cancellationToken);

    public override WebSocketCloseStatus? CloseStatus => _inner.CloseStatus;

    public override string CloseStatusDescription => _inner.CloseStatusDescription;

    public override WebSocketState State => _inner.State;

    public override string SubProtocol => _inner.SubProtocol;

    public override void Abort() => _inner.Abort();

    public override Task CloseAsync(WebSocketCloseStatus closeStatus, string statusDescription, CancellationToken cancellationToken)
        => _inner.CloseAsync(closeStatus, statusDescription, cancellationToken);

    public override Task CloseOutputAsync(WebSocketCloseStatus closeStatus, string statusDescription, CancellationToken cancellationToken)
        => _inner.CloseOutputAsync(closeStatus, statusDescription, _cancellationToken);

    public override void Dispose() => _inner.Dispose();

    public override Task<WebSocketReceiveResult> ReceiveAsync(ArraySegment<byte> buffer, CancellationToken cancellationToken)
        => _inner.ReceiveAsync(buffer, cancellationToken);

    public override Task SendAsync(ArraySegment<byte> buffer, WebSocketMessageType messageType, bool endOfMessage, CancellationToken cancellationToken)
        => _inner.SendAsync(buffer, messageType, endOfMessage, _cancellationToken);
}
```

**Comment by dannyhaak** [comment]
Yes, this works :D.

CTS is canceled when it says 'Cancel now' in the log.

```
[20:07:00.598 DBG HubConnection] Sending StreamItemMessage message '4' completed.
[20:07:00.600 VRB HubConnection] Releasing Connection Lock in SendWithLock (/_/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:957).
[20:07:00.605 VRB HubConnection] Sending item for stream '4'.
[20:07:01.068 VRB HubConnection] Waiting on Connection Lock in HandleConnectionClose (/_/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:1362).
[20:07:01.078 DBG HttpConnection] Disposing HttpConnection.
[20:07:01.085 VRB HubConnection] Stream '4' has been canceled by client.
[20:07:01.092 VRB HubConnection] Sending completion message for stream '4'.
[20:07:01.082 VRB HubConnection] Stream '3' has been canceled by client.
[20:07:01.098 VRB HubConnection] Sending completion message for stream '3'.
[20:07:01.099 VRB HubConnection] Waiting on Connection Lock in CommonStreaming (/_/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:2077).
[20:07:01.087 INF WebSocketsTransport] Transport is stopping.
[20:07:01.101 VRB HubConnection] Waiting on Connection Lock in CommonStreaming (/_/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:2077).
[20:07:01.108 VRB HubConnection] Stream '2' has been canceled by client.
[20:07:01.117 VRB HubConnection] Sending completion message for stream '2'.
[20:07:01.120 VRB HubConnection] Waiting on Connection Lock in CommonStreaming (/_/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:2077).
[20:07:11.262 DBG BackendService] Current cpuUsage 15.6 % and memoryUsage 76836864
[20:07:41.267 DBG BackendService] Current cpuUsage 13.4 % and memoryUsage 75399168
[20:08:11.256 DBG BackendService] Current cpuUsage 13.9 % and memoryUsage 75407360
[20:08:36.308 INF BackendService] Cancel now
[20:08:36.321 INF BackendService] Canceled
[20:08:36.377 DBG WebSocketsTransport] Receive loop canceled.
[20:08:36.381 DBG WebSocketsTransport] Receive loop stopped.
[20:08:36.434 ERR WebSocketsTransport] Error while sending a message.
System.Threading.Tasks.TaskCanceledException: The operation was canceled.
 ---> System.IO.IOException: Unable to write data to the transport connection: Broken pipe.
 ---> System.Net.Sockets.SocketException (32): Broken pipe
   --- End of inner exception stack trace ---
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.ThrowException(SocketError error, CancellationToken cancellationToken)
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.System.Threading.Tasks.Sources.IValueTaskSource.GetResult(Int16 token)
   at System.Net.Security.SslStream.<WriteSingleChunk>g__CompleteWriteAsync|182_1[TIOAdapter](ValueTask writeTask, Byte[] bufferToReturn)
   at System.Net.Security.SslStream.WriteAsyncInternal[TIOAdapter](TIOAdapter writeAdapter, ReadOnlyMemory`1 buffer)
   at System.Net.Http.HttpConnection.RawConnectionStream.WaitWithConnectionCancellationAsync(ValueTask task, HttpConnection connection, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
   at System.Net.Http.HttpConnection.RawConnectionStream.WaitWithConnectionCancellationAsync(ValueTask task, HttpConnection connection, CancellationToken cancellationToken)
   at System.Net.WebSockets.ManagedWebSocket.SendFrameFallbackAsync(MessageOpcode opcode, Boolean endOfMessage, Boolean disableCompression, ReadOnlyMemory`1 payloadBuffer, Task lockTask, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Http.Connections.Client.Internal.WebSocketsTransport.StartSending(WebSocket socket)
[20:08:37.634 DBG WebSocketsTransport] Send loop stopped.
[20:08:37.638 DBG WebSocketsTransport] Transport stopped.
[20:08:37.642 INF HttpConnection] HttpConnection Disposed.
[20:08:37.645 DBG HubConnection] Canceling all outstanding invocations.
[20:08:37.648 ERR HubConnection] Connection is shutting down due to an error.
System.TimeoutException: Server timeout (30000.00ms) elapsed without receiving a message from the server.
[20:08:37.664 VRB HubConnection] The HubConnection is attempting to transition from the Connected state to the Disconnected state.
[20:08:37.669 VRB HubConnection] Releasing Connection Lock in HandleConnectionClose (/_/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:1396).
[20:08:37.682 VRB HubConnection] Releasing Connection Lock in CommonStreaming (/_/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:2081).
[20:08:37.690 VRB HubConnection] Releasing Connection Lock in CommonStreaming (/_/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:2081).
[20:08:37.699 VRB HubConnection] Releasing Connection Lock in CommonStreaming (/_/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:2081).
[20:08:37.707 DBG HubConnection] Invoking the Closed event handler.
[20:08:37.710 INF BackendService] Connection closed: Server timeout (30000.00ms) elapsed without receiving a message from the server.
```

**Comment by dannyhaak** [comment]
Do you believe a temporary work-around might be to prevent SignalR from using Websockets as transport mechanism?

**Comment by BrennanConroy** [comment]
Awesome, thanks for confirming. We have a good idea of what to change in the WebSocketsTransport now to better handle this scenario 😃 

> Do you believe a temporary work-around might be to prevent SignalR from using Websockets as transport mechanism?

Yes that is a temporary work-around. Both SSE and LongPolling should work better in this scenario.

**Comment by dannyhaak** [comment]
Thanks. If this can be solved, I am a happy person.

**Comment by davidfowl** [comment]
W00t excellent! This will make the client more reliable than before 😄

**Comment by BrennanConroy** [comment]
Fixed in 7.0, will be backported to 6.0, likely released in October.

**Comment by dannyhaak** [comment]
Awesome. Thanks a lot for the support and quick turn-around. Much better than I could hope for; continue the great work on dotnet!
